src/gleam/iterator.gleam

```diff
@@ -322,17 +322,23 @@ pub fn append(to first: Iterator(a), suffix second: Iterator(a)) -> Iterator(a)
   |> Iterator
 }
 
-fn do_flatten(continuation: fn() -> Action(Iterator(a))) -> fn() -> Action(a) {
-  fn() {
-    case continuation() {
-      Continue(e, continuation) ->
-        do_append(e.continuation, do_flatten(continuation))()
-      Stop -> Stop
-    }
+fn do_flatten(
+  flattened: fn() -> Action(Iterator(a)),
+  next: fn() -> Action(a),
+) -> Action(a) {
+  case next() {
+    Stop ->
+      case flattened() {
+        Stop -> Stop
+        Continue(it, next_iterator) ->
+          do_flatten(next_iterator, it.continuation)
+      }
+    Continue(e, next_elem) ->
+      Continue(e, fn() { do_flatten(flattened, next_elem) })
```

Comment:
I would prefer to build upon `do_append`, there's no problem having two related functions share private implemention functions. I would generally recommend that

Reply:
We could then make both `do_append` and `do_flatten` return `Action` for readability